x11/xauth.go

```diff
@@ -0,0 +1,164 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package x11
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+)
+
+// See https://cgit.freedesktop.org/xorg/lib/libXau/tree/AuRead.c and
+// https://cgit.freedesktop.org/xorg/lib/libXau/tree/include/X11/Xauth.h
+// for details about the actual file format.
+type xauth struct {
+	Family  uint16
+	Address []byte
+	Number  []byte
+	Name    []byte
+	Data    []byte
+}
+
+func readBytes(f *os.File, data []byte) error {
+	n, err := f.Read(data)
+	if err != nil {
+		return err
+	}
+
+	if n != len(data) {
+		return fmt.Errorf("Could not read enough bytes")
+	}
+
+	return nil
+}
+
+func readChunk(f *os.File) ([]byte, error) {
+	// A chunk consists of a length encoded by two bytes and
+	// additional data which is the real value of the item
+	// we reading here from the file.
+
+	b := [2]byte{}
+	if err := readBytes(f, b[:]); err != nil {
+		return nil, err
+	}
+
+	size := int(binary.BigEndian.Uint16(b[:]))
+	chunk := make([]byte, size)
```

Comment:
This should do some trivial validation of that size before attempting to allocate memory after it.